src/transactions/test/ClaimableBalanceTests.cpp

```diff
@@ -0,0 +1,1030 @@
+
+    int64_t reserve = claimants.size() * lm.getLastReserve();
+    REQUIRE(createAccNativeBeforeCreate - (isNative ? amount : 0) - reserve -
+                static_cast<int64_t>(lm.getLastTxFee()) ==
+            createAccNativeAfterCreate);
+
+    // the claim account doesn't change on the create
+    REQUIRE(claimAccBalanceBeforeCreate == claimAccBalanceAfterCreate);
+
+    if (mergeCreateAcc)
+    {
+        auto root = TestAccount::createRoot(app);
+        createAcc.merge(root);
+    }
+
+    claimAcc.claimClaimableBalance(balanceID);
+
+    if (!mergeCreateAcc)
+    {
+        // check that reserve is sent back to createAcc
+        auto createAccNativeAfterClaim = createAcc.getBalance();
+        REQUIRE(createAccNativeAfterCreate + reserve ==
+                createAccNativeAfterClaim);
+    }
+
+    int64_t fee = isNative ? lm.getLastTxFee() : 0;
+    reserve = mergeCreateAcc ? reserve : 0;
+
+    auto claimAccBalanceAfterClaim = getAccAssetBalance(claimAcc);
+    REQUIRE(claimAccBalanceAfterCreate + amount - fee + reserve ==
+            claimAccBalanceAfterClaim);
+}
+
+TEST_CASE("claimableBalance", "[tx][managedata]")
+{
+    Config const& cfg = getTestConfig();
+
+    VirtualClock clock;
+    auto app = createTestApplication(clock, cfg);
+
+    app->start();
+
+    // set up world
+    auto root = TestAccount::createRoot(*app);
+    auto const& lm = app->getLedgerManager();
+
+    const int64_t trustLineLimit = INT64_MAX;
+    const int64_t minBalance1 = lm.getLastMinBalance(1);
+    const int64_t minBalance3 = lm.getLastMinBalance(3);
+
+    auto acc1 = root.create("acc1", minBalance3);
+    auto acc2 = root.create("acc2", minBalance3);
+
+    auto issuer = root.create("issuer", minBalance3);
+    auto usd = makeAsset(issuer, "USD");
+    auto native = makeNativeAsset();
+
+    // acc2 is authorized
+    acc2.changeTrust(usd, trustLineLimit);
+
+    auto simplePred = makeSimplePredicate(3); // validPredicate
+
+    xdr::xvector<Claimant, 10> validClaimants{makeClaimant(acc2, simplePred)};
+
+    SECTION("not supported before version 14")
+    {
+        for_versions_to(13, *app, [&] {
+            REQUIRE_THROWS_AS(
+                acc1.createClaimableBalance(native, 100, validClaimants),
+                ex_opNOT_SUPPORTED);
+
+            REQUIRE_THROWS_AS(acc1.claimClaimableBalance(ClaimableBalanceID{}),
+                              ex_opNOT_SUPPORTED);
+        });
+    }
+    // generalize for native vs non-native assets
+    for_versions_from(14, *app, [&] {
+        auto balanceTestByAsset = [&](Asset const& asset, int64_t amount) {
+            bool const isNative = asset.type() == ASSET_TYPE_NATIVE;
+
+            // close ledger to change closeTime from 0 so we can test predicates
+            closeLedgerOn(*app, lm.getLastClosedLedgerNum() + 1, 1, 1, 2016);
+
+            // authorize trustline
+            if (!isNative)
+            {
+                SECTION("no trust")
```

Comment:
Does this all actually work as you intend? The sequential sections with intermittent code is a pretty confusing pattern. It leaves me wondering, what _is supposed_ to happen here and what _actually does_ happen here?

Are these early sections like "no trust" and "not authorized" even required? The code in those sections should have no effect other than paying fees and consuming sequence numbers, so the sections might not really be doing anything.

In summary, I'm confused by this.

Reply:
I cleaned up the test cases and rebased into 656f3c9718fa0382f3b983b810179d689ae8e874.

Reply:
I cleaned up the tests and squashed with 572598aa0be8290908319e5d1d98e83c09c5520f.